x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/parser/AbstractBuilder.java

```diff
@@ -137,50 +133,77 @@ public static String unquoteString(Source source) {
-                    replacement = "\\\\";
-                    break;
-                default:
-                    // unknown escape sequence, pass through as-is
-                    replacement = group;
+        StringBuilder sb = new StringBuilder();
+
+        for (int i = 0; i < text.length();) {
+            if (text.charAt(i) == '\\') {
+                // ANTLR4 Grammar guarantees there is always a character after the `\`
+                switch (text.charAt(++i)) {
+                    case 't':
+                        sb.append('\t');
+                        break;
+                    case 'b':
+                        sb.append('\b');
+                        break;
+                    case 'f':
+                        sb.append('\f');
+                        break;
+                    case 'n':
+                        sb.append('\n');
+                        break;
+                    case 'r':
+                        sb.append('\r');
+                        break;
+                    case '"':
+                        sb.append('\"');
+                        break;
+                    case '\'':
+                        sb.append('\'');
+                        break;
+                    case 'u':
+                        i = handleUnicodeChars(source, sb, text, ++i);
+                        break;
+                    case '\\':
+                        sb.append('\\');
+                        // will be interpreted as regex, so we have to escape it
+                        break;
+                    default:
+                        // unknown escape sequence, pass through as-is, e.g: `...\w...`
+                        sb.append('\\').append(text.charAt(i));
+                }
+                i++;
+            } else {
+                sb.append(text.charAt(i++));
             }
+        }
+        return sb.toString();
+    }
 
-            regexMatcher.appendReplacement(resultString, replacement);
-
+    private static int handleUnicodeChars(Source source, StringBuilder sb, String text, int i) {
+        String unicodeSequence;
+        int startIdx = i + 1;
+        int endIdx = text.indexOf('}', startIdx);
+        unicodeSequence = text.substring(startIdx, endIdx);
+        int length = unicodeSequence.length();
+        if (length < 2 || length > 8) {
+            throw new ParsingException(source, "Unicode sequence in curly braces should use [2-8] hex digits, [{}] has [{}]",
+                    text.substring(startIdx - 3, endIdx + 1), length);
         }
-        regexMatcher.appendTail(resultString);
+        sb.append(hexToUnicode(source, unicodeSequence));
+        return endIdx;
+    }
 
-        return resultString.toString();
+    private static String hexToUnicode(Source source, String hex) {
+        int code = Integer.parseInt(hex, 16);
+        // U+D800—U+DFFF can only be used as surrogate pairs and therefore are not valid character codes
+        if (code >= 55296 && code <= 57343) {
```

Comment:
Use the hex declaration when doing the comparison:

`if (code >= 0xD800 && code <= 0xDFFF) ...`